internal/locate/region_request.go

```diff
@@ -2462,3 +2447,59 @@ func (s *replicaSelector) recordAttemptedTime(duration time.Duration) {
 		proxyReplica.attemptedTime += duration
 	}
 }
+
+// canFastRetry returns true if the request can be sent to next replica.
+func (s *replicaSelector) canFastRetry() bool {
+	accessLeader, ok := s.state.(*accessKnownLeader)
+	if ok && accessLeader.isCandidate(s.replicas[accessLeader.leaderIdx]) {
+		// If leader is still candidate, the request will be sent to leader again,
+		// so don't skip since the leader is still busy.
+		return false
+	}
+	return true
+}
+
+type backoffArgs struct {
+	cfg *retry.Config
+	err error
+}
+
+// addPendingBackoff adds pending backoff for the store.
+func (s *replicaSelector) addPendingBackoff(store *Store, cfg *retry.Config, err error) {
+	storeId := uint64(0)
+	if store != nil {
+		storeId = store.storeID
+	}
```

Comment:
I don't know, since there is a `ctx != nil && ctx.Store != nil` check in the up layer https://github.com/tikv/client-go/blob/ea5a5c4ba674f42a468510f4fa4168dad77196af/internal/locate/region_request.go#L1286. And with this check, look like `store` could be nil. But as I known, It should never be nil, I didn't delete this check because I was afraid that there would be a situation that I didn't know.